_posts/2016-10-25-cve-2016-5195-dirtycow-mitigated.md

```diff
@@ -0,0 +1,47 @@
+
+However, not just any old write worked here. In order to trigger the race condition, the process had to write in a way that calls the kernel's `get_user_pages()` function with the `force` parameter set to `1`. The `force` parameter says: "If this page is mapped copy-on-write, then let me write to it (making a private copy) even if the page's protection mode is read-only." As it turns out, it is possible for a memory mapping to be both read-only and copy-on-write, and in fact this is the mode that is usually used when mapping in a program's main binary and shared libraries. Normally, no copy is ever performed, because the writes that would trigger them are not allowed. However, there is a special case where this combination of flags matters: If you are running a program in a debugger, and you ask the debugger to insert a breakpoint, it does so by overwriting the instruction at the given address with a break instruction. That is, it modifies the mapped executable. The `force` flag actually exists for exactly this purpose: so that the debugger can inject breakpoints into the program being executed by the process being debugged (without affecting any other processes that happen to be running the same program).
+
+Because the `force` flag is only useful in very specific circumstances, only certain code paths can trigger the vulnerability. Kernel security engineer and Sandstorm contributor Andrew Lutomirski tells us the only entry points appear to be:
+
+* The `ptrace()` system call's `PTRACE_POKEDATA` operation, which is explicitly meant to be used by debuggers, often for the purpose of setting breakpoints.
+* Writes to `/proc/<pid>/mem`. This appears to be a mistake -- there's no apparent reason why this should use `force`. Probably, someone was copy-pasting code.
+* Various drivers, which are also probably using the flag by mistake.
+
+As it turns out, none of these code paths can be exploited by Sandstorm apps:
+
+* Sandstorm uses seccomp to block the app from invoking `ptrace()`.
+* Sandstorm does not mount `/proc` inside app sandboxes.
+* Sandstorm does not expose driver interfaces from inside app sandboxes. For example, `/dev` contains only `null`, `zero`, `urandom`, and `random` (which is actually a synonym for `urandom`, [as it](https://gist.github.com/tarcieri/6347417) [should be](http://www.2uo.de/myths-about-urandom/)).
```

Comment:
The "as it" "should be" here is excessively cutesy imho but not a showstopper. IMHO unnecessary "see how I'm always right?" language  makes a significant fraction of readers think we're arrogant, which would be a tragic outcome for readers of this blog post.

Not a showstopper; this is a subjective feeling.